### PR TITLE
Remove bad metadata bits cleanup on aarch64

### PR DESCRIPTION
**Pull request by xiangzhai**
Hi @fisk 

I cleanup the code about remove bad metadata bits for aarch64.

I debug the code on aarch64 machine which support `BIC` as equivalent to x86_64 `VPANDN` Logical AND NOT instruction:

```
(gdb) x/22i $pc-44
   0xffffe01e8748:      sub     x1, x1, #0x20
   0xffffe01e874c:      cmp     x15, #0x8
   0xffffe01e8750:      b.ge    0xffffe01e8760  // b.tcont
   0xffffe01e8754:      dcps1   #0xdeae
   0xffffe01e8758:      .inst   0xf7848dc0 ; undefined
   0xffffe01e875c:      .inst   0x0000ffff ; undefined
   0xffffe01e8760:      ldp     q0, q1, [x0, #32]
   0xffffe01e8764:      and     v6.16b, v0.16b, v11.16b
   0xffffe01e8768:      umaxv   b6, v6.16b
   0xffffe01e876c:      fcmp    d6, #0.0
   0xffffe01e8770:      b.ne    0xffffe01e8780  // b.any
=> 0xffffe01e8774:      brk     #0x0
   0xffffe01e8778:      bic     v0.16b, v0.16b, v9.16b
   0xffffe01e877c:      b       0xffffe01e8908
   0xffffe01e8780:      mov     x9, v0.d[0]
   0xffffe01e8784:      ldr     x8, [x28, #48]
   0xffffe01e8788:      tst     x9, x8
   0xffffe01e878c:      b.eq    0xffffe01e8840  // b.none
   0xffffe01e8790:      stp     x29, x30, [sp, #-16]!
   0xffffe01e8794:      mov     x29, sp
   0xffffe01e8798:      stp     x0, x1, [sp, #-144]!
   0xffffe01e879c:      stp     x2, x3, [sp, #16]
(gdb) i r v0 v9
v0             {d = {f = {0x0, 0x0}, u = {0x80000825bd0e510, 0x80000825d08e510}, s = {0x80000825bd0e510, 0x80000825d08e510}}, s = {f = {0x0, 0x0, 0x0, 0x0}, u = {0x5bd0e510, 0x8000082, 0x5d08e510, 0x8000082}, s = {0x5bd0e510, 0x8000082, 0x5d08e510, 0x8000082}}, h = {u = {0xe510, 0x5bd0, 0x82, 0x800, 0xe510, 0x5d08, 0x82, 0x800}, s = {0xe510, 0x5bd0, 0x82, 0x800, 0xe510, 0x5d08, 0x82, 0x800}}, b = {u = {0x10, 0xe5, 0xd0, 0x5b, 0x82, 0x0, 0x0, 0x8, 0x10, 0xe5, 0x8, 0x5d, 0x82, 0x0, 0x0, 0x8}, s = {0x10, 0xe5, 0xd0, 0x5b, 0x82, 0x0, 0x0, 0x8, 0x10, 0xe5, 0x8, 0x5d, 0x82, 0x0, 0x0, 0x8}}, q = {u = {0x80000825d08e510080000825bd0e510}, s = {0x80000825d08e510080000825bd0e510}}}
v9             {d = {f = {0x0, 0x0}, u = {0x1ae0, 0x1ae0}, s = {0x1ae0, 0x1ae0}}, s = {f = {0x0, 0x0, 0x0, 0x0}, u = {0x1ae0, 0x0, 0x1ae0, 0x0}, s = {0x1ae0, 0x0, 0x1ae0, 0x0}}, h = {u = {0x1ae0, 0x0, 0x0, 0x0, 0x1ae0, 0x0, 0x0, 0x0}, s = {0x1ae0, 0x0, 0x0, 0x0, 0x1ae0, 0x0, 0x0, 0x0}}, b = {u = {0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, s = {0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}}, q = {u = {0x1ae00000000000001ae0}, s = {0x1ae00000000000001ae0}}}
(gdb) set $pc+=4
(gdb) si
0x0000ffffe01e877c in ?? ()
(gdb) i r v0 v9
v0             {d = {f = {0x0, 0x0}, u = {0x80000825bd0e510, 0x80000825d08e510}, s = {0x80000825bd0e510, 0x80000825d08e510}}, s = {f = {0x0, 0x0, 0x0, 0x0}, u = {0x5bd0e510, 0x8000082, 0x5d08e510, 0x8000082}, s = {0x5bd0e510, 0x8000082, 0x5d08e510, 0x8000082}}, h = {u = {0xe510, 0x5bd0, 0x82, 0x800, 0xe510, 0x5d08, 0x82, 0x800}, s = {0xe510, 0x5bd0, 0x82, 0x800, 0xe510, 0x5d08, 0x82, 0x800}}, b = {u = {0x10, 0xe5, 0xd0, 0x5b, 0x82, 0x0, 0x0, 0x8, 0x10, 0xe5, 0x8, 0x5d, 0x82, 0x0, 0x0, 0x8}, s = {0x10, 0xe5, 0xd0, 0x5b, 0x82, 0x0, 0x0, 0x8, 0x10, 0xe5, 0x8, 0x5d, 0x82, 0x0, 0x0, 0x8}}, q = {u = {0x80000825d08e510080000825bd0e510}, s = {0x80000825d08e510080000825bd0e510}}}
v9             {d = {f = {0x0, 0x0}, u = {0x1ae0, 0x1ae0}, s = {0x1ae0, 0x1ae0}}, s = {f = {0x0, 0x0, 0x0, 0x0}, u = {0x1ae0, 0x0, 0x1ae0, 0x0}, s = {0x1ae0, 0x0, 0x1ae0, 0x0}}, h = {u = {0x1ae0, 0x0, 0x0, 0x0, 0x1ae0, 0x0, 0x0, 0x0}, s = {0x1ae0, 0x0, 0x0, 0x0, 0x1ae0, 0x0, 0x0, 0x0}}, b = {u = {0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, s = {0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xe0, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}}, q = {u = {0x1ae00000000000001ae0}, s = {0x1ae00000000000001ae0}}}
```

aarch64 fastdebug hotspot:tier1:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/hotspot/jtreg:tier1                     1688  1651    32     5 <<
==============================
```

Please review my patch!

Thanks,
Leslie Zhai

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/zgc pull/15/head:pull/15` \
`$ git checkout pull/15`

Update a local copy of the PR: \
`$ git checkout pull/15` \
`$ git pull https://git.openjdk.org/zgc pull/15/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15`

View PR using the GUI difftool: \
`$ git pr show -t 15`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/zgc/pull/15.diff">https://git.openjdk.org/zgc/pull/15.diff</a>

</details>

### CODE REVIEW

**Comment by fisk** [comment]
Hi @xiangzhai. Is this an attempt to make the arraycopy code look more like x86? What is kind of important is that it's clear whether the responsibility for removing store bad bits, is consistent between loads/stores both for neon and non-neon variants. It looks like you now move the responsibility of removing store bad bits from the store side to the load side, for the neon variant. However, the non-neon variant which is used as a fall-back by the neon code, has the old responsibility where loads are *not* responsible for removing store bad bits, and instead doing that at the store side. This anomaly I think is a bug. When a neon copy load barrier runs and finds one oop to be load-bad, it will use the non-neon copy load barriers as a fallback, and they are not removing any store bad bits - only if a pointer is load bad. So you can have 4 oops, one of them was load bad (which is what triggered the fallback using non-neon instructions). But one of the oops could be load good, but store bad. For example, a marking bit could be wrong, but it has the right remap bits. With this patch, we won't remove the store bad bits from the pointer. Then the 4 oops are combined into a neon register, which eventually gets stored. And the neon store will just or in the store good colors, which was only valid because we had ensured there are no store bad bits. As a result, we will write store bad pointers to the destination, which can crash the JVM.
In summary, I'm okay with moving the store bad uncoloring from the store to the load, but it has to be consistent across non-neon and neon variations of the copy barriers, or we will crash. Without that consistency, this won't work.

**Comment by xiangzhai** [comment]
Hi @fisk 

Thanks for your review!

> Is this an attempt to make the arraycopy code look more like x86?

Yes.

> but it has to be consistent across non-neon and neon variations of the copy barriers

Thanks for pointing out my falut.  And I fixed the add good colors and store new values issue for copy_store_at:

```
static void copy_store_barrier_done(...) {
  // Add good colors
  __ orr(ref, Assembler::T16B, ref, z_copy_store_good_vreg);
  // Store new values
  __ umov(tmp1, ref, Assembler::D, 0);  
  __ umov(tmp2, ref, Assembler::D, 1);
  __ stp(tmp1, tmp2, src);
}

static void copy_store_barrier(...) {
...
  __ fcmpd(vec_tmp, 0.0);
  __ br(Assembler::NE, fallback);

=>  copy_store_barrier_done(masm, new_ref, src, tmp1, tmp2);
  __ b(done);
...
}

void ZBarrierSetAssembler::copy_store_at(...) {
...
  if (is_dest_uninitialized) {
    if (bytes == 32) {
=>      copy_store_barrier_done(masm, src1, Address(dst.base(), dst.offset() + 0), tmp1, tmp2);
=>      copy_store_barrier_done(masm, src2, Address(dst.base(), dst.offset() + 16), tmp1, tmp2);
    } else {
      ShouldNotReachHere();
    }    
  } else {
    // Load pre values
    BarrierSetAssembler::copy_load_at(masm, decorators, type, bytes, vec_tmp1, vec_tmp2, dst, noreg, noreg, fnoreg);

    // Store barrier pre values and color new values
    if (bytes == 32) {
      copy_store_barrier(masm, vec_tmp1, src1, Address(dst.base(), dst.offset() + 0), tmp1, tmp2, tmp3, vec_tmp3);
      copy_store_barrier(masm, vec_tmp2, src2, Address(dst.base(), dst.offset() + 16), tmp1, tmp2, tmp3, vec_tmp3);
    } else {
      ShouldNotReachHere();
    }    
  }
```

aarch64 fastdebug hotspot:tier1:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/hotspot/jtreg:tier1                     1688  1654    34     0 <<
==============================
```

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
aarch64 fastdebug hotspot:tier1 with `VM_OPTIONS=-XX:+UseZGC -XX:-UseSIMDForMemoryOps;`:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/hotspot/jtreg:tier1                     1688  1656    32     0 <<
==============================
```

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 1]
```diff
@@ -545,12 +555,19 @@ static void copy_store_barrier(MacroAssembler* masm,
...
+                                    FloatRegister ref,
+                                    Address src,
+                                    Register tmp1,
+                                    Register tmp2) {
+  // Add good colors
+  __ orr(ref, Assembler::T16B, ref, z_copy_store_good_vreg);
+  // Store new values
+  __ umov(tmp1, ref, Assembler::D, 0);
+  __ umov(tmp2, ref, Assembler::D, 1);
+  __ stp(tmp1, tmp2, src);
```
Comment (fisk) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:574]: Why is there a store here? The barriers should not perform the actual stores. This seems inconsistent with the other code.

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 2]
```diff
@@ -680,17 +705,7 @@ void ZBarrierSetAssembler::copy_store_at(MacroAssembler* masm,
...
-    if (bytes == 8) {
-      __ bfi(src1, tmp1, 0, 16);
-    } else if (bytes == 16) {
-      __ bfi(src1, tmp1, 0, 16);
-      __ bfi(src2, tmp1, 0, 16);
-    } else {
-      ShouldNotReachHere();
-    }
-  } else {
+  if (!is_dest_uninitialized) {
```
Comment (fisk) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:712]: Why did you flip this condition so it's not consistent with the other copy_store_at?

**Review by fisk — COMMENTED** [review]
Some questions...

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp (hunk 1 above)
Comment (xiangzhai) [reply src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:574]: I made the code look too much like x86...

x86 `ZBarrierSetAssembler::copy_store_at`:

```
if (done) {
  color source
=>  store source
} else if (fallback) {
  vector to general register
  copy_store_at for general register
}
```

aarch64:

Remove `Store new values` in the `ZBarrierSetAssembler::copy_store_at`: 

```
-
-  // Store new values
-  BarrierSetAssembler::copy_store_at(masm, decorators, type, bytes, dst, src1, src2, noreg, noreg, noreg, fnoreg, fnoreg, fnoreg);
```

But put the store into `copy_store_barrier_done`, then modified `ZBarrierSetAssembler::copy_store_at`:

```
if (is_dest_uninitialized) {
  copy_store_barrier_done
} else {
  copy_store_barrier
}
```

Because `if (done) else if (fallback)` is in the `copy_store_barrier` for aarch64, so I put a store here:

```
if (done) {
  copy_store_barrier_done
} else if (fallback) {
  vector to general
  copy_store_barrier for general
}
```

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp (hunk 2 above)
Comment (xiangzhai) [reply src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:712]: Ditto, too much like x86...

x86:

```
if (!dest_uninitialized) {
}

color source
store source
```

aarch64:

I added `copy_store_barrier_done` to color and store.

```
if (!is_dest_uninitialized) {
}

=>copy_store_barrier_done
```

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp (hunk 1 above)
Comment (fisk) [reply src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:574]: Okay. It confuses me that there are now two functions called "copy_store_barrier_done", which suggests to me that the responsibility of said functions are similar. But they are completely different - one performs a store, and the other does not. It makes it difficult to read, and to convince myself that we are not storing to the same location twice.

Comment (xiangzhai) [reply src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:574]: Sorry that I am not good at naming functions or my children. I will update my patch to make it better readability.

**Comment by xiangzhai** [comment]
PING @fisk

**Comment by fisk** [comment]
Looks good, thanks for fixing the patch.

**Review by fisk — APPROVED** [review]

**Comment by xiangzhai** [comment]
/integrate


Hi @fisk 

It is my pleasure to contribute for zgc_generational! Could you sponsor to integrate the patch?

Thanks,
Leslie Zhai

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp [hunk 3]
```diff
@@ -779,15 +801,12 @@ void ZBarrierSetAssembler::copy_store_at(MacroAssembler* masm,
 
     // Store barrier pre values and color new values
     if (bytes == 32) {
-      copy_store_barrier(masm, vec_tmp1, src1, Address(dst.base(), dst.offset() + 0), tmp1, tmp2, tmp3, vec_tmp3);
+      copy_store_barrier(masm, vec_tmp1, src1, Address(dst.base(), dst.offset() + 0),  tmp1, tmp2, tmp3, vec_tmp3);
```
Comment (fisk) [inline src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:808]: Since you removed the store below that used to perform the actual store, the copy_store_barrier function kind of needs to perform the store instead, or we will lose one of the stores. But reading that function, it looks like it only performs the store *sometimes*. In the fast path color_and_store_source is used in the fast path, which does indeed perform the store. But when the slow path is used, the "other" copy_store_barrier for normal registers is used, but that one does not perform the store. So when we go through the slow path, the stores are lost and we effectively don't copy as intended.

Now the intended contract which is shared across all GC barrier code across all architectures, is that the barriers never have the responsibility of performing the accesses. This bug demonstrates why - because the responsibilities need to be clear or we get bugs. The store below should be re-instated, and the copy_store_barrier should never perform the actual store.

**Review by fisk — CHANGES_REQUESTED** [review]
I looked again and found another bug.

File: src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp (hunk 3 above)
Comment (xiangzhai) [reply src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp:808]: Hi @fisk 

Thanks for your careful review!

x86_64 perform the store in the slow-path:

```
void ZBarrierSetAssembler::copy_store_at(MacroAssembler* masm,
                                         DecoratorSet decorators,
                                         BasicType type,
                                         size_t bytes,
                                         Address dst,
                                         XMMRegister src,
                                         Register tmp1,
                                         Register tmp2,
                                         XMMRegister xmm_tmp) {
...
  } else if (bytes == 32) {
    // 256 bits vector

    if (!dest_uninitialized) {
      Load destination vector
      Check destination store_bad_mask
      __ jcc(Assembler::notZero, fallback);
    }    

 =>   Color and store source // fast-path
    __ jmp(done);

 =>   __ bind(fallback); // slow-path

    __ subptr(rsp, wordSize * 4);
    Store 256 bits vector register in the stack

    // The lower 64 bits.
    Load tmp1 from stack disp 0
=>    copy_store_at dst0 disp 0  // Perform the store in the slow-path

    // The mid-lower 64 bits.
    Load tmp1 from stack disp 8
    copy_store_at dst1 disp 8

    // The mid-higher 64 bits.
    Load tmp1 from stack disp 16
    copy_store_at dst2 disp 16

    // The higher 64 bits.
    Load tmp1 from stack disp 24
    copy_store_at dst3 disp 24

    __ addptr(rsp, wordSize * 4);

    __ bind(done);
  }
...
```

Sorry ... aarch64 forgot to perform the store in the slow-path:

```
static void copy_store_barrier(MacroAssembler* masm,
                               FloatRegister pre_ref,
                               FloatRegister new_ref,
                               Address src, 
                               Register tmp1,
                               Register tmp2,
                               Register tmp3,
                               FloatRegister vec_tmp) {
  Label done, fallback;

  Test reference against bad mask
  __ br(Assembler::NE, fallback);

=>  Color and store source // fast-path
  __ b(done);

=>  __ bind(fallback); // slow-path

  // Extract the 2 oops from the pre_ref vector register
  Unsigned Move vector element The lower 64 bits to general-purpose register
=>  copy_store_barrier(masm, tmp2, noreg, Address(src.base(), src.offset() + 0), tmp1, tmp3); // NOT perform the actual store in the slow-path

  Unsigned Move vector element The higher 64 bits to general-purpose register
  copy_store_barrier(masm, tmp2, noreg, Address(src.base(), src.offset() + 8), tmp1, tmp3);

  __ bind(done);
}
```

> The store below should be re-instated, and the copy_store_barrier should never perform the actual store.

How about `if (new_ref == noreg)` then perform the actual store?

> Now the intended contract which is shared across all GC barrier code across all architectures, is that the barriers never have the responsibility of performing the accesses. This bug demonstrates why - because the responsibilities need to be clear or we get bugs.

I debugged [x86_64 on purpose forgot to store value in the slow-path](https://github.com/xiangzhai/zgc/commit/1a1dd3a95aaa383da487c2ed89a7849bbd82bc1c) then jtreg testcase `compiler/arraycopy/stress/TestStressArrayCopy.java` reproduced the issue which I brought in:

```
STDERR:
 stdout: [For random generator using seed: 8894202274802264953
To re-run test with same seed value please add "-Djdk.test.lib.random.seed=8894202274802264953" to command line.
];
 stderr: [Exception in thread "main" java.lang.RuntimeException: Error after seed: 10 elements, at pos 3
        at compiler.arraycopy.stress.AbstractStressArrayCopy.throwSeedError(AbstractStressArrayCopy.java:54)
        at compiler.arraycopy.stress.StressObjectArrayCopy.testWith(StressObjectArrayCopy.java:44)
        at compiler.arraycopy.stress.AbstractStressArrayCopy.exhaustiveWith(AbstractStressArrayCopy.java:105)
        at compiler.arraycopy.stress.AbstractStressArrayCopy.run(AbstractStressArrayCopy.java:146)
        at compiler.arraycopy.stress.StressObjectArrayCopy.main(StressObjectArrayCopy.java:82)
]
 exitValue = 1

java.lang.RuntimeException: Expected to get exit value of [0], exit value is: [1]
        at jdk.test.lib.process.OutputAnalyzer.shouldHaveExitValue(OutputAnalyzer.java:490)
        at compiler.arraycopy.stress.TestStressArrayCopy.main(TestStressArrayCopy.java:188)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
        at java.base/java.lang.Thread.run(Thread.java:1623)
```

So there is no `src/hotspot/share/gc/z` related bug found.

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
PING @fisk
